apps/src/applab/dataBrowser/DataTable.jsx

```diff
@@ -29,6 +41,75 @@ const DataTable = React.createClass({
     onViewChange: React.PropTypes.func.isRequired
   },
 
+  getInitialState() {
+    return {
+      newColumns: [],
+      editingColumn: null
+    };
+  },
+
+  componentWillReceiveProps(nextProps) {
+    // Forget about new columns when switching between tables.
+    if (this.props.tableName !== nextProps.tableName) {
+      this.setState({newColumns: []});
+    }
+  },
+
+  /**
+   * @param {Array|undefined} newColumns
+   */
+  addColumn() {
+    const columnName = this.getNextColumnName();
+    this.setState({
+      newColumns: this.state.newColumns.concat(columnName),
+      editingColumn: columnName,
+    });
+  },
+
+  deleteColumn(columnToRemove) {
+    if (confirm('Are you sure you want to delete this entire column? You cannot undo this action.')) {
+      this.setState({
+        newColumns: this.state.newColumns.filter(column => column !== columnToRemove)
+      });
+      FirebaseStorage.deleteColumn(
+        this.props.tableName,
+        columnToRemove,
+        () => {},
+        error => console.warn(error)
+      );
+    }
+  },
+
+  /**
+   * @param {string|null} columnName Column to edit, or null to not edit any column.
+   */
+  editColumn(columnName) {
+    this.setState({editingColumn: columnName});
+  },
+
+  renameColumn(oldName, newName) {
+    let newColumns = this.state.newColumns.map(
+      curName => (curName === oldName ? newName : curName)
+    );
+    // Append the new name if the old name was not found.
+    if (newColumns.indexOf(newName) === -1) {
+      newColumns.push(newName);
+    }
+    this.setState({
+      newColumns,
+      editingColumn: null
+    });
+    if (oldName !== newName) {
+      FirebaseStorage.renameColumn(
+        this.props.tableName,
+        oldName,
+        newName,
+        () => {},
```

Comment:
The empty success callback here makes me wonder if we're missing an intermediate state.  Should the column name field go into some kind of locked/saving state before we call Firebase, and only change to 'not editing' when the save is complete?